File: src/main/java/api/routes/requestchannel/dto/ToggleRequestChannelButtonDto.java
package api.routes.requestchannel.dto;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class ToggleRequestChannelButtonDto {
    @NotBlank()
    String server_id;
    @NotBlank()
    @Pattern(regexp = "^(previous|rewind|stop|pnp|skip|favourite|loop|shuffle|disconnect|filters)$", flags = { Pattern.Flag.CASE_INSENSITIVE})
    String button;
}

File: src/main/kotlin/api/routes/requestchannel/dto/ToggleRequestChannelButtonDto.kt
package api.routes.requestchannel.dto

import io.ktor.server.plugins.requestvalidation.*
import kotlinx.serialization.Serializable
import main.utils.GeneralUtils.isDiscordId

@Serializable
data class ToggleRequestChannelButtonDto(
    val server_id: String,
    val button: String
)

fun RequestValidationConfig.validateToggleRequestChannelButtonDto() {
    validate<ToggleRequestChannelButtonDto> { dto ->
        if (!dto.server_id.isDiscordId())
            return@validate ValidationResult.Invalid("The server id must be a valid Discord id!")

        when (dto.button.lowercase()) {
            "previous", "rewind", "stop",
            "pnp", "skip", "favourite", "loop",
            "shuffle", "disconnect", "filters" -> ValidationResult.Valid

            else -> ValidationResult.Invalid("${dto.button} is an invalid button!")
        }
    }
}
